### ISSUE TITLE
PTRHash fails to build on 10^12 keys

### ISSUE THREAD

**Issue by vigna** [RagnarGrootKoerkamp/PTRHash#6]
```
[...]
part    507 bucket  0.01%  chain:   8388608
Too many displacements. Aborting!
Possible causes:
- Too many elements in part.
- Not enough empty slots => lower alpha.
- Not enough buckets     => increase c.
- Not enough entropy     => fix algorithm.

part   1350 bucket  0.03%  chain:   8388608
Too many displacements. Aborting!
Possible causes:
- Too many elements in part.
- Not enough empty slots => lower alpha.
- Not enough buckets     => increase c.
- Not enough entropy     => fix algorithm.

thread 'main' panicked at /home/vigna/git/PTRHash/src/lib.rs:369:13:
Failed to find a global seed after 3 tries for 262144 keys.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I can send the whole log if that's useful.

**Comment by vigna** [comment]
BTW, somewhat unrelated, but this comment
```
    // Since running all queries is super slow, we only check a subset of them.
    // Although this doesn't completely check that there are no duplicate
    // mappings, by the birthday paradox we can be quite sure there are none
    // since we check way more than sqrt(n) of them.
````
has no mathematical basis. It would be correct if the output of the function was _random_, but it is not. If there are exactly two keys colliding the probability of hitting them is the probability of having exactly those two keys in the subset you test, and this has nothing to do with the birthday paradox.

**Comment by RagnarGrootKoerkamp** [comment]
Yeah so as the error says, the algorithm only works well if there are sufficiently many buckets and slots. You probably want to run with `c=10` or `c=11` to have more buckets and significantly increase the probability that displacing finds a solution.

I still have to work out the math at some point. Currently the number of buckets scales as c * n / lg(n) (as does PTHash), but in practice it seems that the default c=9 fails for larger n, so it's needed to compensate manually for the /lg(n) by increasing c.

**Comment by vigna** [comment]
That was 10, I'll try 11.

**Comment by RagnarGrootKoerkamp** [comment]
Did this end up working?

**Comment by vigna** [comment]
I tried with c=11 and a single shard but I got
```
        keys: 1000000000000
      shards:          1
       parts:    3892549
   slots/prt:     262144
   slots tot: 1020408365056
 buckets/prt:      72285
 buckets tot: 281372904465
 keys/bucket:          3.55
memory allocation of 2147483648 bytes failed
```

Now trying with more shards.

**Comment by vigna** [comment]
With c=11 the first round failed. I'll make it do another couple of rounds. I'm using 1000 shards.

**Comment by RagnarGrootKoerkamp** [comment]
Then c=12 maybe?

**Comment by RagnarGrootKoerkamp** [comment]
Or maybe try alpha=0.97 or a bit lower. Its hard to know exactly where is the bottleneck.

**Comment by vigna** [comment]
It didn't stop, but it is still at 800/1000 chunks after three days. Hardly usable in practice. If it finishes I'll try with a smaller alpha.

**Comment by RagnarGrootKoerkamp** [comment]
Hmm. That's weird. Did you enable sharding to disk? Without it it will hash all keys once for each shard which is quite a bit slower probably.

**Comment by vigna** [comment]
Oh no. Well, from my experience, it is better that default options offer a reasonable behavior. People won't always come to you for explanations—they'll just move elsewhere.

**Comment by vigna** [comment]
On a better note: it completed! But it didn't print any stats about space, unless " keys/bucket:          3.26" it is. That's where serialization is really useful—you can measure the file to double check.

**Comment by vigna** [comment]
Will it obey to TMPDIR? I can't find where it's creating the shards...

**Comment by RagnarGrootKoerkamp** [comment]
Yeah well... People using this for up to 10^9 elements probably don't want sharding to disk and this program writing disk unexpectedly. 
Maybe I'll make a separate function build_on_disk for visibility and clarity.
TMPDIR env var should work yes.

It should normally print a bits/element stat at the end, but at some point I broke that and it was printing 0 instead. I think that's fixed now.

**Comment by vigna** [comment]
Well, you might just log what you're doing, since you're already logging anyway. Something like "100 shards, keys will be re-read 100 times (switch to offline sharding for single-pass)". At least people will know what's happening and they can take action.

I agree on the "unexpectedly on disk" viewpoint—VFunc does the same. But in the VFunc case if things do not go well you'll have an out-of-memory error, and people will look for offline options because that's what happens in 90% of these structures. The problem with PTRHash is that you do not get out-of-memory—you get slow. People does not take this behavior as a suggestion to change options, since it somehow works.

I modified the example to serialize (it'll be fun). I'll keep you posted. Probably with memory-mapping I can even try to validate the structure.

**Comment by vigna** [comment]
> Yeah well... People using this for up to 10^9 elements probably don't want sharding to disk and this program writing disk unexpectedly. Maybe I'll make a separate function build_on_disk for visibility and clarity. TMPDIR env var should work yes.
> 

It does!
```
TMP PATH: "/mnt/extra/analysis/vigna/.tmpnLSHXt"
```

**Comment by RagnarGrootKoerkamp** [comment]
> Well, you might just log what you're doing, since you're already logging anyway. Something like "100 shards, keys will be re-read 100 times (switch to offline sharding for single-pass)". At least people will know what's happening and they can take action.

Good idea. I will add some logs.

> I agree on the "unexpectedly on disk" viewpoint—VFunc does the same. But in the VFunc case if things do not go well you'll have an out-of-memory error, and people will look for offline options because that's what happens in 90% of these structures. The problem with PTRHash is that you do not get out-of-memory—you get slow. People does not take this behavior as a suggestion to change options, since it somehow works.

Hmm very interesting point. So it sounds like going out-of-memory if a preferred default behaviour over both sharding with rereading and sharding to disk. I'll probably just make 3 distinct entry points for these options.

**Comment by vigna** [comment]
> Hmm very interesting point. So it sounds like going out-of-memory if a preferred default behaviour over both sharding with rereading and sharding to disk. I'll probably just make 3 distinct entry points for these options.

"Is" is a big word :). That's my personal take on user psychology. Catastrophic disasters lead to search for solutions. Bad performance leads to think the that's the performance of the software. Sometimes the user will think "maybe I can fiddle", but I think less often than with a catastrophic result.

**Comment by RagnarGrootKoerkamp** [comment]
Allright, went over things and cleaned up the docs in `src/lib.rs` a bit. Can you run off the git version here? Then I'll push it to crates.io as 2.0.0 once things work on you end.

Some notes:

**hash function:**
You said you use string keys right? Does `GxHash` work for you? It uses AES encryption hardware simd intrinsics, which makes it very fast and much less branchy, but this requires somewhat modern hardware. Otherwise just use `Xxh3`.
You should use the 128-bit variant probably.

**bucket-fn:**
`PtrHashParams::default()` now just uses the `default_fast` config from the paper, with linear bucket fn.

**remap:**
By default I have `Vec<u32>`, but with 10^12 ~ 2^40 keys, that's not gonna work. `CacheLineEF` is also only made up to 40-bit keys, and anyway is IMO slightly too brittle to use here. There is a small probability that construction fails on having the remap 'gaps' too far spread out for CacheLineEF to handle. So: you'll want to use either `Vec<u64>` which doubles the remap size and brings the total to ~3.35 bits/key, or you can use your EliasFano, which is ~2.8 instead, but will give slightly slower queries. (A compact vec would also work of course, but I don't think I have that supported yet.) Either way I could also consider adding some support for translating from one remap structure to another.
string keys: The A

**Comment by rahul-iyer** [comment]
@RagnarGrootKoerkamp Any idea when will the 2.0.0 be released to crates.io?

**Comment by RagnarGrootKoerkamp** [comment]
Ah I was waiting for a message here that current master works. I could also push a new version now, but would be nice if you could test it first.
But if it's much easier for you to depend on crates.io, I can maybe push an alpha/beta version there and then you can depend on that instead of this repo.

**Comment by RagnarGrootKoerkamp** [comment]
@rahul-iyer I meanwhile pushed `2.0.0-alpha.1` to crates.io. You could pin the cargo.toml to that specific version and I think 10^12 keys should just work with the new default configuration.

**Comment by RagnarGrootKoerkamp** [comment]
(going to close this; things should be better now.)
